# phaseLD
Developed by Alexandre Marand & Hanain Zhao, 2017

A simple LD based method to quickly phase pseudo-oneway test cross populations by identifying window haplotypes via Bayesian methods (within phaseLD.pl). Optionally, users can implement an HMM on the raw SNP haplotype calls, which models error and recombination rates between loci to infer the most probable haplotypes.

Please cite "" if you use this software.  

## Dependencies
phaseLD.pl requires users to have the following perl modules available to @INC:

1) [Sort::Naturally](http://search.cpan.org/~bingos/Sort-Naturally-1.03/lib/Sort/Naturally.pm)
2) [Pod::Usage](http://search.cpan.org/~marekr/Pod-Usage-1.69/lib/Pod/Usage.pm)
3) [List::Util](http://search.cpan.org/~pevans/Scalar-List-Utils-1.47/lib/List/Util.pm)
4) [Parallel::ForkManager](search.cpan.org/~yanick/Parallel-ForkMâ€¦)

All modules are available from CPAN, and can be accessed with cpan/cpanm.

Click [here] for link to cpanm.

[here]: http://search.cpan.org/~miyagawa/Menlo-1.9004/script/cpanm-menlo

## Usage
```
Name:
    F1 Diploid Genotype Phasing, Alexandre Marand & Hainan Zhao

Usage:
    phaseLD.pl --in <file.gen> [OPTIONS]

Description:
    Below you will find a list of parameter options. The defaults work well
    for clean data: low genotype missing rate, low genotype error rate, and
    high marker density (10,000-1,000,000).

  Input/Output:
    --in|-i             Genotype file [String|REQUIRED]

    --out|-o            Prefix for output files [String|Default='phased']

  Performance:
    --win|-w            Set minimum window size for estimating LD
                        [Int|Default=20]

    --step|-s           Set minimum step size for LD calculations in a
                        sliding window, defaults to 1 [Int|Default=1]

    --bwin|-n           Set minimum window size for bayes haplotype calling
                        [Int|Default=50]

    --bstep|-t          Set minimum step size for bayes haplotype calling
                        [Int|Default=5]

    --rpen|-r           Set minimum r2 value needed to retain SNPs. This
                        threshold helps to remove false positive SNP call.
                        Defaults to 0.25. [Int|Default=0.25]

  Run Time Options:
    --threads|-p        Set number of threads [Int|Default=1]

    --quick_mode|-f     Instead of calculating LD for every marker in the
                        window, calculate LD for window_size/10. The next
                        position is the first marker above the R2 threshold.
                        This makes the computation 10-20X faster for the
                        default window size. Time saved increases
                        exponentially as window sizes get larger.

    --fast_mode|-q      Rather than use pairwise LD calculations within a
                        window, use LD chaining instead. Starting at marker
                        X, the algorithm searches for the nearest marker, y,
                        with R2 above the threshold. The next iteration
                        starts at marker y and continues until all markers
                        are exhausted.

    --filter|-f         Optionally use file with suffix 'bad' generated by
                        phaseLD to filter poor markers. 'prefix.bad'
                        contains IDs of bad SNPs identified in the first
                        pass. Useful to cleaning haplotype switch problems
                        [String|Optional]

  Misc:
    --help|-h           Print a brief help message and exit
```
